processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/UnsafeSortDataRows.java

```diff
@@ -140,19 +143,25 @@ public void initialize() throws MemoryException, CarbonSortKeyAndGroupByExceptio
     semaphore = new Semaphore(parameters.getNumberOfCores());
   }
 
-  private UnsafeCarbonRowPage createUnsafeRowPage()
-      throws MemoryException, CarbonSortKeyAndGroupByException {
-    MemoryBlock baseBlock =
-        UnsafeMemoryManager.allocateMemoryWithRetry(this.taskId, inMemoryChunkSize);
-    boolean isMemoryAvailable =
-        UnsafeSortMemoryManager.INSTANCE.isMemoryAvailable(baseBlock.size());
-    if (isMemoryAvailable) {
-      UnsafeSortMemoryManager.INSTANCE.allocateDummyMemory(baseBlock.size());
-    } else {
-      // merge and spill in-memory pages to disk if memory is not enough
-      unsafeInMemoryIntermediateFileMerger.tryTriggerInmemoryMerging(true);
+  private UnsafeCarbonRowPage createUnsafeRowPage() {
+    try {
+      MemoryBlock baseBlock =
+          UnsafeMemoryManager.allocateMemoryWithRetry(this.taskId, inMemoryChunkSize);
+      boolean isMemoryAvailable =
+          UnsafeSortMemoryManager.INSTANCE.isMemoryAvailable(baseBlock.size());
+      if (isMemoryAvailable) {
+        UnsafeSortMemoryManager.INSTANCE.allocateDummyMemory(baseBlock.size());
+      } else {
+        // merge and spill in-memory pages to disk if memory is not enough
+        unsafeInMemoryIntermediateFileMerger.tryTriggerInmemoryMerging(true);
+      }
+      return new UnsafeCarbonRowPage(tableFieldStat, baseBlock, !isMemoryAvailable, taskId);
+    } catch (MemoryException | CarbonSortKeyAndGroupByException e) {
+      // This will set rowPage reference to null. If not set, other threads will use same reference.
+      // As handlePreviousPage() free the rowPage.
+      // If not set to null, rowPage will be accessed again after free by other thread.
+      return null;
```

Comment:
I think it is better to throw MemoryException to call so caller no need to check whether it is null

Reply:
Issue came because of this itself. Throwing exception will not set rowPage reference to null. So, other thread will access this rowPage. But rowPage was already freed from previous thread. hence jvm crash